docs/spec/lightclient/attacks/evidence-handling.md

```diff
@@ -0,0 +1,250 @@
+
+
+#### Helper functions
+
+We assume the following helper functions:
+
+```go
+// Returns trace of verified light blocks starting from rootHeight and ending with targetHeight.
+trace(lightStore LightStore, rootHeight int64, targetHeight int64) LightBlock[]
+
+
+
+```
+
+
+```go
+func DetectLightClientAttacks(primary PeerID, 
+                              primary_trace []LightBlock, 
+                              witness PeerID) (LightClientAttackEvidence, LightClientAttackEvidence) {
+    primary_lca_evidence, witness_trace = DetectLightClientAttack(primary_trace, witness)
+    
+    witness_lca_evidence = nil
+    if witness_trace != nil {
+        witness_lca_evidence, _ = DetectLightClientAttack(witness_trace, primary)
+    }
+    return primary_lca_evidence, witness_lca_evidence
+}
+
+func DetectLightClientAttack(trace []LightBlock, peer PeerID) (LightClientAttackEvidence, []LightBlock) {
+
+    trusted = trace[0]
+    lightStore = new LightStore().Update(trusted, StateTrusted)
+
+    for i in 1..len(trace)-1 {
+        lightStore, result = VerifyToTarget(peer, lightStore , i) 
+   
+        if result == ResultFailure then return (nil, nil)
+        
+        current = lightStore.Get(i)
+        
+        // if obtained header is the same as in the trace we continue with a next height
+        if current.Header == trace[i].Header continue
+
+        // we have identified a conflicting header
+        attackType = nil
+        trustedBlock = trace[i-1]
+        conflictingBlock = trace[i]
+        
+        if !isPossibleBlock(trustedBlock, conflictingBlock) { 
+            attackType = LunaticAttack
+        } else if trustedBlock.Header.Round == conflictingBlock.Header.Round {
+            attackType = EquivocationAttack
+        } else {
+           attackType = AmnesiaAttack
+        }                                        
+                 
+        return (LightClientAttackEvidence { conflictingBlock, trustedBlock.Header.Height, attackType }, 
+                trace(lightStore, trace[i-1].Height, trace[i].Height))
+    } 
+    return (nil, nil)       
+}
+```
+
+## Evidence handling 
+
+As part of on chain evidence handling, full nodes identifies misbehaving processes and informs
+the application, so they can be slashed. We now specify for each event type evidence handling logic.
+
+### Lunatic attack evidence handling
+
+```go
+func detectMisbehavingProcesses(lcAttackEvidence LightClientAttackEvidence, bc Blockchain) []ValidatorAddress {
+   assume lcAttackEvidence.Type == LunaticAttack
+   misbehavingValidators = []ValidatorAddress
+
+   conflictingHeight = lcAttackEvidence.ConflictingBlock.Header.Height
+   
+   bondedValidators = bc.GetValidators(lcAttackEvidence.CommonHeight) union 
+                      bc.GetNextValidators(lcAttackEvidence.CommonHeight) union
+                      bc.GetValidators(conflictingHeight)) 
```

Comment:
Also for further clarification, this should be the validators in the set that actually voted for the conflicting header not just all validators